client/network/src/discovery.rs

```diff
@@ -137,9 +144,42 @@ impl DiscoveryBehaviour {
 		}
 	}
 
+	/// Enable discovery for the given protocol.
+	pub fn add_discovery(&mut self, p: ProtocolId) {
+		if self.kademlias.contains_key(&p) {
+			warn!(target: "sub-libp2p", "Discovery already registered for protocol {:?}", p);
+			return
+		}
+
+		let mut config = KademliaConfig::default();
+
+		let proto_name =
+			if libp2p::kad::protocol::DEFAULT_PROTO_NAME == p.as_bytes() {
+				// Temporary hack to retain backwards compatibility. Once this version
+				// has been rolled out we should remove the special handling of
+				// DEFAULT_PROTO_NAME and work only with proper protocol IDs.
+				Vec::from(p.as_bytes())
+			} else {
+				let mut v = vec![b'/'];
+				v.extend_from_slice(p.as_bytes());
+				v.extend_from_slice(b"/kad");
+				v
+			};
+		config.set_protocol_name(proto_name);
+
+		let store = MemoryStore::new(self.local_peer_id.clone());
+		let mut kad = Kademlia::with_config(self.local_peer_id.clone(), store, config);
+
+		for (peer_id, addr) in &self.user_defined {
+			kad.add_address(peer_id, addr.clone());
+		}
+
+		self.kademlias.insert(p, kad);
+	}
+
 	/// Returns the list of nodes that we know exist in the network.
 	pub fn known_peers(&mut self) -> impl Iterator<Item = &PeerId> {
-		self.kademlia.kbuckets_entries()
+		self.kademlias.values_mut().map(|k| k.kbuckets_entries()).flatten()
```

Comment:
This is probably going to produce a lot of duplicates.

Reply:
Do you want me to deduplicate here?

Reply:
Well, nevermind, I deduplicate the peers now.